src/java.base/share/classes/java/util/GregorianCalendar.java

```diff
@@ -1307,7 +1307,13 @@ public void roll(int field, int amount) {
                             woy = min;
                         }
                     }
-                    set(field, getRolledValue(woy, amount, min, max));
+                    int newWeekOfYear = getRolledValue(woy, amount, min, max);
+                    // Make sure that the (potential) minimum week has the
+                    // current DAY_OF_WEEK
+                    if (newWeekOfYear == 1 && isInvalidWeek1()) {
+                        newWeekOfYear+=1;
```

Comment:
is `+1` always correct? Does it work when the amount is negative?

Reply:
When a week is rolled down to 1 (and it does not have a valid day_of_week), the existing code would appropriately adjust min, so that the new week set would be 52. However, yes, it should only be incremented if amount is positive, but there does not need to be a case for when amount is negative.